python/postprocessing/modules/btv/btagSFProducer.py

```diff
@@ -193,25 +193,26 @@ def beginFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
     def endFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
         pass
 
-    def getSF(self, pt, eta, flavor, syst = 'central', wp = 'M', measurement_type = 'auto', shape_corr = False, discr = 0.):
-        """Evaluate the SFs.
-            Note the flavor convention: hadronFlavor is b = 5, c = 4, f = 0
-            Convert them to the btagging group convention of 0 ,1, 2
-
-            Same for working points: input is 'L', 'M', 'T'
-            Convert to 0, 1, 2
-
-            Automatically checks if values are in allowed range
-
-            If unknown wp/syst/mtype/flavor, returns -1.0
+    def getReader(self, wp, shape_corr = False):
         """
+            Get btag scale factor reader.
+            Convert working points: input is 'L', 'M', 'T', 'shape_corr' to 0, 1, 2, 3
+        """
+        if shape_corr:
+            wp = "shape_corr"
+        wp_btv = { "l" : 0, "m" : 1, "t" : 2, "shape_corr" : 3 }.get(wp.lower(), None)
+        if wp_btv == None or not wp_btv in self.readers.keys():
+            if self.verbose > 0:
+                print("WARNING: Unknown working point '%s', setting b-tagging SF to -1!" % wp)
+            return -1.
```

Comment:
Not a maintainer of this project, just wanted to point out that if there's no WP available for a particular b-tagging discriminator (e.g. shape corrections are missing for deep CSV at the time of writing), then this function returns a float but in successful cases it returns a reader object. It's going to raise an exception in `getSFs()` if the WP is not available, whereas the current (and IMO preferable) behavior is that we set the SFs to 1 in such cases. You can view and maybe cherry-pick my fix into your branch that is subject to this PR: https://github.com/HEP-KBFI/nanoAOD-tools/commit/174af42506dcbeed1558a9f4a34b4c40e1d2e998

Reply:
That is the same behaviour as is [currently implemented](https://github.com/cms-nanoAOD/nanoAOD-tools/blob/master/python/postprocessing/modules/btv/btagSFProducer.py#L230), is it not?
The goal of this PR was to change the implementation without changing any behaviour.

Also, at no point a reader object is returned. `reader.eval_auto_bounds()` returns a double (see [BTagCalibrationReader.h](https://github.com/cms-sw/cmssw/blob/aeb38655a73431dbf2eeab4e033d15fbb99cf0df/CondTools/BTau/interface/BTagCalibrationReader.h))

Reply:
No, `getSF()` in current implementation *always* returns a float, and the value will be -1 if there's no reader available. The SF is eventually set to 1 if `getSF()` returns too small of a SF:
https://github.com/cms-nanoAOD/nanoAOD-tools/blob/fd69163b283bfcaebe9f6d181db02b41bdf7435a/python/postprocessing/modules/btv/btagSFProducer.py#L277-L280
In the PR you've split up `getSF()` into multiple functions, one part being `getReader()` function which either returns -1 or a reader object ([link](https://github.com/kreczko/nanoAOD-tools/blob/7da623bfebecf3155e5689316fddf7a5c3920d3c/python/postprocessing/modules/btv/btagSFProducer.py#L207-L208)). You get the reader once before looping over systematics ([link](https://github.com/kreczko/nanoAOD-tools/blob/7da623bfebecf3155e5689316fddf7a5c3920d3c/python/postprocessing/modules/btv/btagSFProducer.py#L275)) and then pass it to `getSFs()` in the loop ([link](https://github.com/kreczko/nanoAOD-tools/blob/7da623bfebecf3155e5689316fddf7a5c3920d3c/python/postprocessing/modules/btv/btagSFProducer.py#L278)). Now, the reader could point to either a valid ,,reader'' object if the corrections are available, or to -1 if the corrections are not available. The problem here is that you don't check if the reader passed to `getSFs()` is actually a ,,reader'' object or something else ([here](https://github.com/kreczko/nanoAOD-tools/blob/7da623bfebecf3155e5689316fddf7a5c3920d3c/python/postprocessing/modules/btv/btagSFProducer.py#L241), [here](https://github.com/kreczko/nanoAOD-tools/blob/7da623bfebecf3155e5689316fddf7a5c3920d3c/python/postprocessing/modules/btv/btagSFProducer.py#L243) and [here](https://github.com/kreczko/nanoAOD-tools/blob/7da623bfebecf3155e5689316fddf7a5c3920d3c/python/postprocessing/modules/btv/btagSFProducer.py#L245)). Calling `eval_auto_bounds()` on a float results in a runtime error.

[My solution](https://github.com/HEP-KBFI/nanoAOD-tools/commit/174af42506dcbeed1558a9f4a34b4c40e1d2e998) was to return `None` in `getReader()` if the corrections are missing ([link](https://github.com/HEP-KBFI/nanoAOD-tools/blob/174af42506dcbeed1558a9f4a34b4c40e1d2e998/python/postprocessing/modules/btv/btagSFProducer.py#L207)) so that we won't mix two different data types, and then add a protection to `getSFs()` before calling any functions on the `reader` object ([link](https://github.com/HEP-KBFI/nanoAOD-tools/blob/174af42506dcbeed1558a9f4a34b4c40e1d2e998/python/postprocessing/modules/btv/btagSFProducer.py#L240-L251)).

If it's still confusing, add
```python
btagSF2017_deep = lambda : btagSFProducer("2017", algo = "deepcsv")
```
to your PR and run it.

Reply:
I see. The diff was indeed confusing as there are changes across two functions.
Will make the change.